File: Data Structures & Algorithms/anagram-groups/submission-0.cpp
class Solution {
public:

    bool checkAnagrams(string s1, string s2){
        map<char, int> mp;
        int m = s1.length(), n = s2.length();
        if(m != n)
            return false;
        else{
            for(int i = 0; i < m; i++){
                mp[s1[i]]++;
            }
            for(int j = 0; j < n; j++){
                mp[s2[j]]--;
            }

            for(auto it:mp){
                if(it.second != 0)
                    return false;
            }
            return true;
        }
    }
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        int n = strs.size();
        vector<vector<string>> res;
        vector<string> temp;
        for(int i = 0; i < n; i++){
            string s = strs[i];
            int tmp = 0;
            for(int j = 0; j < res.size(); j++){
                if(checkAnagrams(res[j][0], s) == true){
                    res[j].push_back(s);
                    tmp = 1;
                }
            }
            if(tmp == 0){
                temp.push_back(s);
                res.push_back(temp);
                temp.clear();
            }
        }
        return res;
    }
};

File: Data Structures & Algorithms/task-scheduling/submission-0.java
class Solution {
    public int leastInterval(char[] tasks, int n) {
        int len = tasks.length;
        PriorityQueue<Map.Entry<Character, Integer>> pq = new PriorityQueue<>((a,b) -> b.getValue() - a.getValue());

        Map<Character, Integer> mp = new HashMap<>();
        for(int i = 0; i < len; i++)
            mp.put(tasks[i], mp.getOrDefault(tasks[i], 0) + 1);
        
        pq.addAll(mp.entrySet());

        Map<Character, Integer> temp = new HashMap<>();
        int res = 0;
        while(!pq.isEmpty()){
            int minCycles = 0;
            while(!pq.isEmpty() && !((n+1) == minCycles)){
                Map.Entry<Character,Integer> t = pq.poll();
                if(t.getValue() > 1)
                    temp.put(t.getKey(), t.getValue()-1);
                minCycles++;
            }

            if(temp.size() == 0)
                res += minCycles;
            else if(n+1 == minCycles)
                res += minCycles;
            else 
                res += n+1;
            pq.addAll(temp.entrySet());
            temp.clear();
        }
        return res;
    }
}
